extensions/amp-o2-player/0.1/test/validator-amp-o2-player.html

```diff
@@ -0,0 +1,56 @@
+<!--
+  Copyright 2016 The AMP HTML Authors. All Rights Reserved.
+
+  Licensed under the Apache License, Version 2.0 (the "License");
+  you may not use this file except in compliance with the License.
+  You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+  Unless required by applicable law or agreed to in writing, software
+  distributed under the License is distributed on an "AS-IS" BASIS,
+  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+  See the License for the specific language governing permissions and
+  limitations under the license.
+-->
+<!--
+  Test Description:
+  Tests support for the amp-o2-player tag.
+  https://github.com/ampproject/amphtml/blob/master/extensions/amp-o2-player/amp-o2-player.md
+-->
+<!doctype html>
+<html ⚡>
+<head>
+  <meta charset="utf-8">
+  <link rel="canonical" href="./regular-html-version.html" />
+  <meta name="viewport" content="width=device-width,minimum-scale=1">
+  <style amp-boilerplate>body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}</style><noscript><style amp-boilerplate>body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}</style></noscript>
+  <script async custom-element="amp-o2-player" src="https://cdn.ampproject.org/v0/amp-o2-player-latest.js"></script>
+  <script async src="https://cdn.ampproject.org/v0.js"></script>
+</head>
+<body>
+  <!-- Example of a valid amp-o2-player -->
+  <amp-o2-player
+      data-pid="573d9bf1e4b02a3388f42c36"
+      data-bcid="50d595ec0364e95588c77bd2"
+      layout="responsive" width="480" height="270">
+  </amp-o2-player>
+
+  <!-- Valid: data-video-id, data-player-id are optional;
+       leaving them out results in the default config. -->
+  <amp-o2-player
+      data-pid="573d9bf1e4b02a3388f42c36"
+      data-bcid="50d595ec0364e95588c77bd2"
+      data-bid="573d9bd3e4b00df34e4325ed"
+      data-vid="57029f58e4b09a9cae6cc82f"
+      data-macros="m.playertype=HTML5"
+      layout="responsive" width="480" height="270">
+  </amp-o2-player>
+
+  <!-- Invalid: atleast vid or pid & bcid are needed;
+       leaving them out results in an error. -->
```

Comment:
Your comment here suggests that this case should generate an error, but the test case passes. There isn't exactly support for rules of the form (A or (B and C)), but you can get the same result. 

I think what you can do is in the `.protoscii', something like this:
```
  attrs: {
    name: "data-vid"
    mandatory_oneof: "data-vid or data-pid and data-bcid"
  }
  attrs: {
    name: "data-bcid"
    mandatory_oneof: "data-vid or data-pid and data-bcid"
    trigger: {
      also_requires_attr: "data-pid"
    }
  }
```

This does prevent a tag from having all 3 though as the mandatory_oneof feature is an 'exactly one' check. Do you want to have tags with all three (data-vid, data-pid, and data-bcid)?